dali/kernels/scratch.h

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_SCRATCH_H_
+#define DALI_KERNELS_SCRATCH_H_
+
+#include "context.h"
+#include <array>
+#include <utility>
+#include <type_traits>
+
+namespace dali {
+
+template <typename T>
+class BumpAllocator {
+  static_assert(std::is_trivially_constructible<T>::value && std::is_trivially_destructible<T>::value,
+    "BumpAllocator must be used with trivially constructible and destructible objects");
+ public:
+  BumpAllocator(T *mem = nullptr, size_t total = 0) : memory_(mem), total_(total), used_(0) {}
+  BumpAllocator(BumpAllocator &&a) : BumpAllocator() {
+    *this = std::move(a);
+  }
+
+  ~BumpAllocator() {
+    for (size_t i = 0; i < used_; i++) {
+      memory_[i].~T();
+    }
+  }
+
+  inline BumpAllocator &operator=(BumpAllocator &&a) {
+    std::swap(memory_, a.memory_);
+    std::swap(total_, a.total_);
+    std::swap(used_, a.used);
+    return *this;
+  }
+
+  inline T *New(size_t elements) {
+    EnforceAvail(elements);
+    T *p = memory_ + used_;
+    used_ += elements;
+    return p;
+  }
+
+  inline T *next() const { return memory + used_; }
+
+  inline size_t avail() const { return used_ - total_; }
+  inline size_t total() const { return total_; }
+  inline size_t used() const { return used_; }
+
+  inline void EnforceAvail(size_t required) {
+    DALI_ENFORCE(used_ + elements <= total_);
+  }
+
+ private:
+  T *memory_;
+  size_t total_;
+  size_t used_;
+};
+
+class Scratchpad : public ScratchpadAllocator {
+ public:
+  Scratchpad(std::array<BumpAllocator<char>, (size_t)StorageBackend::Count> &&allocs) : allocs_(std::move(allocs)) {}
+
+  void *Alloc(StorageBackend backend, size_t bytes, size_t alignment) override {
+    auto &A = allocs_[(int)backend];
+    size_t ptr = (size_t)A.next();
+    // Calculate padding needed to satisfy alignmnent requirements
+    size_t padding = (alignment-1) & -ptr;
+    A.New(padding);
+    return A.New(bytes);
+  }
+ private:
+  std::array<BumpAllocator<char>, (size_t)StorageBackend::Count> allocs_;
```

Comment:
Why not:
```suggestion
  std::array<BumpAllocator<char>, static_cast<size_t>(StorageBackend::Count)> allocs_;
```

Reply:
If that's really a question "why": it's too long for what it does _here_.
`static_cast` is invaluable in the context of pointer/reference casting, where it ensures that the types are actually related (and can actually change the numeric value of pointers!) - or in case of templates where it will raise alarm if someone supplies a type that cannot be safely cast, loses qualifiers, etc. When you know that you're casting an enumeration to an integral type, `static_cast<>` is just adding 13 characters to the line length.
TL;DR
So, if I'm to satisfy the linter, I'd prefer constructor-style cast: `size_t(StorageBackend::Count)`.